Gems/Atom/Utils/Code/Include/Atom/Utils/StableDynamicArray.inl

```diff
@@ -676,6 +675,8 @@ namespace AZ
         if (m_data)
         {
             m_page = other.m_page;
+            m_weakHandle = StableDynamicArrayWeakHandle<ValueType>{ other.m_weakHandle.m_pageIndex, other.m_weakHandle.m_elementIndex };
+
```

Comment:
I realize this whole function is existing behavior, but it seems weird that if I try to assign `handle1 = handle2;` and they are of unrelated types it deletes both of them.

Any chance we can update it to test the cast before doing anything and not free things in the error case?